pkg/lightning/backend/tidb/tidb.go

```diff
@@ -411,27 +479,81 @@ func (be *tidbBackend) WriteRowsToDB(ctx context.Context, tableName string, colu
 		insertStmt.WriteByte(')')
 	}
 	insertStmt.WriteString(" VALUES")
+	return &insertStmt
+}
 
-	// Note: we are not going to do interpolation (prepared statements) to avoid
-	// complication arise from data length overflow of BIT and BINARY columns
+// Only used for TestWriteRowsErrorSkip
+var mockErrorCount = 0
 
-	for i, row := range rows {
-		if i != 0 {
-			insertStmt.WriteByte(',')
-		}
-		insertStmt.WriteString(string(row))
-	}
+func (be *tidbBackend) execStmts(ctx context.Context, stmtTasks []stmtTask, batch bool) error {
+stmtTaskLoop:
+	for _, stmtTask := range stmtTasks {
+	retryLoop:
+		for i := 0; i < writeRowsMaxRetryTimes; i++ {
+			stmt := stmtTask.stmt
+			txnCtx, cancel := context.WithCancel(ctx)
+			defer cancel()
+			tx, err := be.db.BeginTx(txnCtx, &sql.TxOptions{})
+			if err != nil {
+				if !common.IsContextCanceledError(err) {
+					log.L().Error("begin a transaction failed before executing statement",
+						zap.Int("row_index_start", stmtTask.rowIdxStart), zap.Int("row_index_end", stmtTask.rowIdxEnd),
+						zap.Array("rows", stmtTask.rows), zap.String("stmt", redact.String(stmt)), zap.Error(err))
+				}
+				goto errCheck
+			}
+			_, err = tx.ExecContext(txnCtx, stmt)
+
+			failpoint.Inject("mockNonRetryableError", func() {
+				if batch || stmtTask.rowIdxStart == 1 || stmtTask.rowIdxStart == 3 {
```

Comment:
Shall we depend on gomock to return this kind of error instead of depending on this kind of some what weird ways?

Reply:
I haven't any better idea to do this yet, but some comments are added. PTAL.